spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaListenerAnnotationBeanPostProcessor.java

```diff
@@ -407,18 +407,32 @@ public Object postProcessAfterInitialization(final Object bean, final String bea
 					this.logger.debug(() -> annotatedMethods.size() + " @KafkaListener methods processed on bean '"
 							+ beanName + "': " + annotatedMethods);
 				}
+				Set<Method> methodsWithHandler = MethodIntrospector.selectMethods(targetClass,
+						(ReflectionUtils.MethodFilter) method ->
+								AnnotationUtils.findAnnotation(method, KafkaHandler.class) != null);
+				boolean hasMethodLevelKafkaHandlerAnnotation = !methodsWithHandler.isEmpty();
 				if (hasClassLevelListeners) {
-					Set<Method> methodsWithHandler = MethodIntrospector.selectMethods(targetClass,
-							(ReflectionUtils.MethodFilter) method ->
-									AnnotationUtils.findAnnotation(method, KafkaHandler.class) != null);
 					List<Method> multiMethods = new ArrayList<>(methodsWithHandler);
 					processMultiMethodListeners(classLevelListeners, multiMethods, targetClass, bean, beanName);
 				}
+				throwErrorIfNoListenerMethods(bean, hasMethodLevelListeners,
+											hasClassLevelListeners, hasMethodLevelKafkaHandlerAnnotation);
 			}
 		}
 		return bean;
 	}
 
+	private void throwErrorIfNoListenerMethods(Object bean, boolean hasMethodLevelListeners,
+				boolean hasClassLevelListeners, boolean hasMethodLevelKafkaHandlerAnnotation) {
+		if (hasMethodLevelListeners) {
+			return;
+		}
+
+		if (hasClassLevelListeners && !hasMethodLevelKafkaHandlerAnnotation) {
+			throw new IllegalStateException("No kafka listener methods found on bean type: " + bean.getClass());
```

Comment:
I think just `on bean` is enough, without any `getClass()` extraction.
The bean instance might lead to the place where it declared.